src/Universum/List/Safe.hs

```diff
@@ -103,3 +109,22 @@ maximumBy = F.maximumBy
 maximum :: Ord a => NonEmpty a -> a
 maximum = F.maximum
 {-# INLINE maximum #-}
+
+-- | Operates like 'groupByFst', but uses the provided getters
+-- for the key to group by and the value.
+--
+-- >>> groupByKey (`mod` 5) id [1, 6, 7, 2, 12, 1]
+-- [1 :| [6], 7 :| [2, 12], 1 :| []]
+groupByKey :: Eq k => (a -> k) -> (a -> v) -> [a] -> [(k, NonEmpty v)]
+groupByKey toKey toVal l =
+  map
+    (\ne -> (toKey (NE.head ne), NE.map toVal ne))
+    (NE.groupWith toKey l)
```

Comment:
If there isn't enough inlining into and around this function, then it could leak keys into the values. I think it's best to do something like this instead:

```haskell
groupByKeyWith :: (Eq k', Foldable f) => (k -> k') -> (a -> (k, v)) -> f a -> [(k, NonEmpty v)]
```

I can try to write something up.

Reply:
Here you go:

```haskell
groupByKeyWith :: (Eq k', Foldable f) => (k -> k') -> (a -> (k, v)) -> f a -> [(k, NonEmpty v)]
groupByKeyWith kt split = go . Foldable.toList
  where
    go [] = []
    go (a : as)
      | (k, v) <- split a
      , let (ys, zs) = more (kt k) as
      = (k, v :| ys) : zs

    more _ [] = ([], [])
    more ko' (a : as)
      | (kn, v) <- split a
      , let kn' = kt kn
      = if ko' == kn'
         then let rec = more ko' as
                 in (v : fst rec, snd rec)
         else let (vs, rec) = more kn' as
                 in ([], (kn, v :| vs) : rec)

groupByKey :: (Eq k, Foldable f) => (a -> (k, v)) -> f a -> [(k, NonEmpty v)]
groupByKey = groupByKeyWith id

groupByFst :: (Eq a, Foldable f) => f (a, b) -> [(a, NonEmpty b)]
groupByFst = groupByKey id
```

Reply:
My claim may not have been quite precise. You wrote

```haskell
groupByKey toKey toVal l =
  map
    (\ne -> (toKey (NE.head ne), NE.map toVal ne))
    (NE.groupWith toKey l)
```

Suppose we're just looking at pairs (if `toKey` is expensive, then your version also has the downside of calculating it twice for some values). The `groupWith` call gives you something like

```haskell
[ (k1, v11) :| [(k1', v12), ...]
, (k2, v21) :| [(k2', v22), ...]
, ...]
```

The prime marks signify values that are `==` but not pointer equal (these are common).

Let's look at just the first element of the result. The function mapping over it gives us

```haskell
let ne = (k1, v11) :| [(k1', v12), ...]
in (toKey (NE.head ne), NE.map toVal ne)
```

The first component will start out a thunk, but fortunately it will probably be a selector thunk, so the garbage collector should be able to clean it up.

The second component will be a regular thunk, which will incrementally walk the list dropping keys. Until that's walked, however, those `k1'`, `k2'`, etc., values will be retained. Even where the keys are being dropped, we don't do it directly but rather create selector thunks for the GC to clean up. This is only true with optimization (which is fine), because `snd` needs to inline into `map` to make a selector thunk rather than a function application thunk.

So the risk of a serious memory leak is probably low, but it makes a lot of sense that things are kind of slow.

Reply:
Yeah, that sounds reasonable, thanks!

And you propose the function with extra `(k -> k')` indirection because you know the use cases when this `(k -> k')` argument cannot be filled with just `id`, or because that may help GC to cleanup the keys earlier in some way?

Reply:
Just that it's often useful to do things like that. I could easily imagine sticking something like `Data.Text.toLower` in there. It's also quite possible to use a more general `k -> k -> Ordering` for a `groupBy`-like thing. Could add that too.

Reply:
Yes, `Bool`. Sorry.

Reply:
Thanks again! I changed the impl according to your suggestion, could you review the result?

One minor difference: I went with universum's `Container` instead of `Foldable`, because I remember some tickets in this repo dedicated to switching from `Foldable`.